### PR TITLE
Add trailing slashes to REST API paths

### LINKED ISSUE #5084: Work around potential issues with REST API redirects due to trailing slashes
## Feature Description

Tracking ticket for #5004, head over there for more details.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
On some setups, all REST API routes get suffixed with a trailing slash. So `https://example.com/wp-json/web-stories/v1/web-story` will get redirected to `https://example.com/wp-json/web-stories/v1/web-story/`.

This doesn't matter to WordPress as `rest_api_loaded` removes any trailing slashes from the requested route. So requesting `https://example.com/wp-json/web-stories/v1/web-story/////` will be seen as requesting `/wp-json/web-stories/v1/web-story`.

However, the editor expects and operates with non-suffixed routes everywhere. This causes a couple of problems:

1. The request `body` might get dropped for `POST` requests due to the redirect
1. We preload `/wp-json/web-stories/v1/web-story`, but the editor requests `/wp-json/web-stories/v1/web-story/` - the preload does not work because the requests don't match

**Let's fix this.**

The simplest solution is to just append trailing slashes everywhere. However, that breaks the preloading because `rest_preload_api_request` does not remove the trailing slashes in the same way `rest_api_loaded` does. That means preloading `/wp-json/web-stories/v1/web-story/` will throw an error because it does not recognize this route with a trailing slash.

I've reported this at https://core.trac.wordpress.org/ticket/51636.

~To work around this, I've ported `rest_preload_api_request` to our code base and made the necessary adjustment to call `untrailingslashit`.

This way we can use paths with trailing slashes consistently everywhere to support setups like the one mentioned here.~

Edit: we've already ported `rest_preload_api_request` to our code base in the meantie, so this PR just adds the `untrailingslashit` change to it.

**To-do**

* [ ] Add tests to enforce trailing slashes

**User-facing changes**

N/A

**Testing Instructions**

N/A

### CODE REVIEW

File: includes/Dashboard.php [hunk 1]
```diff
@@ -233,7 +233,7 @@ public function load_stories_dashboard() {
 
 		$preload_data = array_reduce(
 			$preload_paths,
-			'rest_preload_api_request',
+			'\Google\Web_Stories\rest_preload_api_request',
```
Comment (spacedmonkey) [inline includes/Dashboard.php]: ?

File: includes/templates/admin/edit-story.php [hunk 2]
```diff
@@ -36,7 +36,7 @@
 // Preload common data.
 $preload_paths = [
 	sprintf( '/web-stories/v1/%s/%s?context=edit', $rest_base, $post->ID ),
```
Comment (spacedmonkey) [inline includes/templates/admin/edit-story.php]:
```suggestion
	sprintf( '/web-stories/v1/%s/%s/?context=edit', $rest_base, $post->ID ),
```
???

File: includes/namespace.php [hunk 3]
```diff
@@ -219,3 +220,77 @@ function get_plugin_instance() {
...
+			$method = 'GET';
+		}
+	}
+
+	$path_parts = wp_parse_url( $path );
+	if ( false === $path_parts ) {
+		return $memo;
+	}
+
+	$request = new WP_REST_Request( $method, untrailingslashit( $path_parts['path'] ) );
```
Comment (spacedmonkey) [inline includes/namespace.php:188]: Is this the only line difference? 

After this line, we could just call
```
return \rest_preload_api_request( $memo, $path );
```

File: includes/Dashboard.php (hunk 1 above)
Comment (swissspidy) [reply includes/Dashboard.php]: it's all explained in the PR description

File: includes/templates/admin/edit-story.php (hunk 2 above)
Comment (swissspidy) [reply includes/templates/admin/edit-story.php]: ah yes, forgot that one

File: includes/namespace.php (hunk 3 above)
Comment (swissspidy) [reply includes/namespace.php:188]: Yes. Good idea to call the original function 👍

Comment (swissspidy) [reply includes/namespace.php:188]: Or well, I'd still need to somehow merge the resulting arrays 🤔

File: assets/src/edit-story/app/api/apiProvider.js [hunk 4]
```diff
@@ -42,7 +43,9 @@ function APIProvider({ children }) {
 
   const getStoryById = useCallback(
     (storyId) => {
-      const path = addQueryArgs(`${stories}/${storyId}`, { context: `edit` });
+      const path = addQueryArgs(`${trailingslashit(stories)}${storyId}`, {
```
Comment (swissspidy) [inline assets/src/edit-story/app/api/apiProvider.js]:
```suggestion
      const path = addQueryArgs(`${trailingslashit(stories)}${storyId}/`, {
```

**Review by spacedmonkey — APPROVED** [review]
👍

**Pull request merged**